lib/logstash/outputs/logstash.rb

```diff
@@ -61,129 +54,189 @@ def initialize(*a)
       fail LogStash::ConfigurationError, 'The `logstash` output does not have an externally-configurable `codec`'
     end
 
-    if @ssl_certificate_authorities && @ssl_certificate_authorities.size > 1
-      fail LogStash::ConfigurationError, 'The `logstash` output supports at most one `ssl_certificate_authorities` path'
+    @headers = {}.tap do | header |
+      header["Content-Type"] = "application/x-ndjson"
+      header["Content-Encoding"] = "gzip"
     end
 
-    logger.debug("initializing inner HTTP output plugin")
-    @internal_http = plugin_factory.output('http').new(inner_http_output_options)
-    logger.debug("inner HTTP output plugin has been initialized")
+    logger.debug("`logstash` output plugin has been initialized.")
   end
 
   def register
-    logger.debug("registering inner HTTP output plugin")
-    @internal_http.register
-    logger.debug("inner HTTP output plugin has been registered")
+    logger.debug("Registering `logstash` output plugin.")
+
+    # TODO: confirm if we have a plan to change @user to @username
+    logger.warn("Both `user` and `username` are defined, using `username` value.") if @user && @username
+    @user = @username ? @username.freeze : @user
+
+    if @ssl_enabled == false
+      rejected_ssl_settings = @original_params.keys.select { |k| k.start_with?('ssl_') } - %w(ssl_enabled)
+      fail(LogStash::ConfigurationError, "Explicit SSL-related settings not supported because `ssl_enabled => false`: #{rejected_ssl_settings}") if rejected_ssl_settings.any?
+    end
+
+    # if we don't initialize now, we get runtime error when sending events if there are issues with configs
+    @http_client = client
+    @load_balancer = LoadBalancer.new(construct_host_uri)
+
+    logger.debug("`logstash` output plugin has been registered.")
   end
 
   def multi_receive(events)
     return if events.empty?
-    logger.trace("proxying #{events.size} events to inner HTTP plugin")
-    @internal_http.multi_receive(events)
-  rescue => e
-    logger.error("inner HTTP plugin has had an unrecoverable exception: #{e.message} at #{e.backtrace.first}")
-    raise
+
+    send_events(events)
   end
 
   def stop
-    logger.debug("stopping inner HTTP output plugin")
-    @internal_http.stop
-    logger.debug('inner HTTP output plugin has been stopped')
+    logger.debug("Stopping `logstash` output plugin.")
+    # TODO: what should we stop?
```

Comment:
@yaauie do you know what we should stop here?

Reply:
Looking at the HTTP output we were previously using, it:

 - [sends `Timer#cancel`](https://github.com/logstash-plugins/logstash-output-http/blob/v5.6.0/lib/logstash/outputs/http.rb#L296) to `@timer` (which presumably should have zero pending tasks).
   - The http output is designed to send batches of events one-at-a-time (with `format => json_batch` hacked on top to send whole batches), and each batch gets its own short-lived blocking `Queue`, working from that queue until it emits a `:done`. A shared `Timer` is used to _delay_ putting retries into the batch's short-lived queue so that other entries from the batch can continue making progress, and must be stopped when the plugin is stopped.
   - because this implementation only ever works with whole batches, we do not need the complexity of a blocking queue with out-of-band retries and can instead operate in a block with a `retry` keyword (or in a loop that exits upon success or shutdown); if we do this, we don't need a timer
 - [sends `Manticore::Client#close`](https://github.com/logstash-plugins/logstash-output-http/blob/v5.6.0/lib/logstash/outputs/http.rb#L297) to the `client` from the http mixin, allowing it to free up resources. Since we are including the http mixin directly, we should do the same.